pyvrp/cpp/Individual.h

```diff
@@ -110,4 +112,21 @@ class Individual
 // Outputs an individual into a given ostream in VRPLIB format
 std::ostream &operator<<(std::ostream &out, Individual const &indiv);
 
+namespace std
+{
+template <> struct hash<Individual>
+{
+    std::size_t operator()(Individual const &individual) const
+    {
+        size_t res = 17;
+        res = res * 31 + std::hash<size_t>()(individual.nbRoutes);
+        res = res * 31 + std::hash<size_t>()(individual.distance);
+        res = res * 31 + std::hash<size_t>()(individual.capacityExcess);
+        res = res * 31 + std::hash<size_t>()(individual.timeWarp);
+
+        return res;
```

Comment:
Question: what happens if the two hash values of an individual happen to be the same? Individuals that have the same but permuted routes all have the hash value right?

Reply:
Objects with the same hash value end up in the same bucket in a hash table (basically: they're put under the same index), and a hash table then has to linearly search through all items in the bucket using `operator==` to compare the items stored there. So typically you want your hash values to spread out, which is the most efficient way to use a hash table (spreading over all the buckets, which limits the amount of linear searches needed). 

For permuted individuals, that's fine. They are equal when compared using `operator==`, so the search stops at the first duplicate.

Reply:
I have two additional questions:
* Why do we need the hash function? I think this is to efficiently identify duplicates among different objects, correct? Where do we do that?
* The hash function changes if the individual changes, which is currently possible (e.g. `LocalSearch::exportIndividual`). What does that mean if this individual is inside some collection using the hash? Is there an implicit contract that individuals should not change once they're in some collection? (which makes sense but is tricky, we already have challenges with cost functions changing with penalties, which you luckily did not use in the hash ;)).

Reply:
Sorry, I had missed this comment yesterday because the conversation was already marked resolved.

>Why do we need the hash function? I think this is to efficiently identify duplicates among different objects, correct? Where do we do that?

In a number of tests, both for mappings (dict keys) and set membership.

>What does that mean if this individual is inside some collection using the hash? Is there an implicit contract that individuals should not change once they're in some collection? 

As a C++ programmer: yeah, that'd a user issue. As a Python programmer: yeah, we should probably make sure that does not happen. I do really like the idea of being able to hash individuals, however, since it's pretty convenient when you need to do a reverse lookup of individuals to their index in tests. 

In #137 and #138 I fix this problem by making the Individual immutable altogether. That also simplifies some of the complications in the GeneticAlgorithm needing to copy the new best solution - it's fairly straightforward now.